[AryanG01] iP

AryanG01 opened the pull request:
### **Axela**

> The tragedy in life doesn’t lie in not reaching your goal. The tragedy lies in having no goal to reach - Benjamin E. Mays [(src)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

**Let all the worries disappear from your mind, because Axela is there for you now**

- Text-based
- Easy to learn
- ~FAST~ **_SUPER_** FAST to use 

All you need to do is,

- download it from [here](https://github.com/AryanG01/ip).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x]  Managing todo tasks, deadlines, and events
- [ ]  GUI (coming soon)
- [ ]  Reminders (coming soon)
- [ ] Having a conversation (coming soon)

****

**You can try to `run` it too! Here's the main method**
```
public static void main(String[] args) {
        new Duke("./src/main/java/duke/duke.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

AryanG01 committed:
Level-0

AryanG01 committed:
Level-1

AryanG01 committed:
Level-2

AryanG01 committed:
Level-3

AryanG01 committed:
Level-4

AryanG01 committed:
A-TextUiTesting

AryanG01 committed:
Level-5

AryanG01 committed:
Level-6

AryanG01 committed:
A-Enums

AryanG01 committed:
A-TextUiTesting

AryanG01 committed:
Level-7

AryanG01 committed:
Level-7

AryanG01 committed:
Level-7

AryanG01 committed:
master

AryanG01 committed:
Merge branch 'master' into branch-Level-7

AryanG01 committed:
Level-8

AryanG01 committed:
Merge branch 'master' into branch-Level-8

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/TaskList.java
#	src/main/java/duke.txt

AryanG01 committed:
Merge branch 'branch-Level-8'

junnengsoo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,51 @@
+        Runtime.getRuntime().addShutdownHook(new Thread(tasks::saveTasksToFile));
+
+        System.out.println("____________________________________________________________");
+        System.out.println(" Hello! I'm Axela");
```
Love the name of your bot!

junnengsoo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,51 @@
+
+        while (true) {
+            String command = scanner.nextLine();
+            System.out.println("____________________________________________________________");
```
Is it possible to abstract all the printing of lines into a variable that can be used throughout your code?

junnengsoo commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,214 @@
+        StringBuilder sb = new StringBuilder();
+        sb.append(" Here are the tasks in your list:\n");
+        if (!tasks.isEmpty()) {
+<<<<<<< HEAD
```
Do remember to resolve your merge conflicts!

junnengsoo reviewed (commented):
Remember to push your code from last week, else amazing work!

AryanG01 committed:
commit changes

AryanG01 committed:
A-MoreOOP

nikele2001 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,127 @@
+            switch (taskType) {
+                case TODO:
+                    // Parse the description, fromDate, and toDate
+                    String[] todoParts = parts[1].split(" /from ");
```
perhaps it would be good to split using strings without the black space characters followed by trimming the elements in todoParts when you want to use them? I find that it might be helpful if the user of the program accidentally enters multiple blankspace characters

nikele2001 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,253 @@
+        // Parse the line and create tasks based on the format in the file
+        String[] parts = line.split(" \\| ");
+        if (parts.length != 3) {
+            throw new DukeException("Invalid task format in the file.");
```
Perhaps subtypes of DukeException can be defined to specify error message printed instead of manually entering the error message every time DukeExceptions are thrown.

nikele2001 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,253 @@
+        if (index >= 0 && index < tasks.size()) {
+            return tasks.get(index);
+        } else {
+            return null;
```
Is there a reason why getTask() do not throw a DukeException specifying that invalid task index was inputted?

nikele2001 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,34 @@
+public class Task {
```
Perhaps the Task class can be abstracted since there will not be the situation where it is instantiated (either Todo, Deadline or Event will be instantiated and not Task itself)

nikele2001 reviewed (commented):
In general, good job! It was great that you considered speeding up string manipulation operations by using the StringBuilder class, and also OOP principles are generally adhered to. However, methods in the Ui class seems to be repetitive and it would be good if you can find a way to abstract these patterns to make the code leaner. Otherwise, well done! :)

AryanG01 committed:
A-Packages

AryanG01 committed:
changes

AryanG01 committed:
A-Gradle

AryanG01 committed:
Revert "A-Gradle"

This reverts commit b3e74b03742fc1ae87fab3db2c7d532dd3941bf8.

AryanG01 committed:
Revert "Revert "A-Gradle""

This reverts commit 1801946d9a31e1d23e474ee658ded3e23394e6d9.

AryanG01 committed:
A-Gradle

AryanG01 committed:
A-JUnit

AryanG01 committed:
A-Jar

AryanG01 committed:
A-JavaDoc

AryanG01 committed:
A-CodingStandard

AryanG01 committed:
Level-9

AryanG01 committed:
change

AryanG01 committed:
Merge branch 'branch-A-JavaDoc'

AryanG01 committed:
A-CodingStandard

AryanG01 committed:
Level-9

AryanG01 committed:
A-CheckStyle

AryanG01 committed:
Level-10

AryanG01 committed:
Level-10

AryanG01 committed:
remove emoji from marking and unmarking already marked and unmarked tasks

AryanG01 committed:
Update duke.txt

AryanG01 committed:
Add 'assert' Statements for Parameter Validation

Enhance code reliability by introducing 'assert' statements to validate
method parameters. 'assert' statements are added at the beginning of
relevant methods to ensure that critical parameters such as 'command'
and 'line' are not null. This helps catch potential issues early in
development, making the code more robust.

Changes Made:
- Add Javadoc header comment for getResponse function in 'Duke' class
- Add 'assert' statements in the 'Ui' class to validate 'task'
  parameters.
- Add 'assert' statements in the 'Parser' class to validate 'command'
  and 'line' parameters.
- Add 'assert' statements in the 'TaskList' class to validate 'task'
  and 'keyword' parameters.
- Convert 'Task' class into an abstract class

These 'assert' statements serve as a form of internal documentation,
clearly indicating the expected parameter conditions and enhancing the
overall maintainability of the codebase.

AryanG01 committed:
Merge pull request #2 from AryanG01/Branch-A-Assertions

Add 'assert' Statements for Parameter Validation

AryanG01 committed:
Added new Command classes and abstracted Command hierarchy

The existing code lacked a Help command to display all available
commands to the user.

To provide users with a way to view all available commands, we
introduced a new HelpCommand class. This class generates a message
listing all the available commands and their descriptions.

In addition to adding the HelpCommand, we abstracted the Command
hierarchy, ensuring that all Command classes extend a common Command
base class. This abstraction allows for more consistent handling of
commands and enhances maintainability.

This approach was chosen because it provides a user-friendly way for
users to learn about the available commands and how to use them,
while also improving the code structure and maintainability.

AryanG01 committed:
Merge pull request #3 from AryanG01/branch-A-CodeQuality

Added new Command classes and abstracted Command hierarchy

AryanG01 committed:
A-CI

AryanG01 committed:
change

AryanG01 committed:
Sorting tasks by priority.

To provide a more organized task list for better task management, the
tasks are being sorted by their priority levels.

The PriorityComparator class is used to compare tasks based on their
priorities. This way, tasks with higher priorities appear first in the
list.

The change enhances task prioritization and improves user experience.

AryanG01 committed:
commit changes

AryanG01 committed:
A-BetterGui

AryanG01 committed:
commit changes

AryanG01 committed:
A-MoreErrorHandling

AryanG01 committed:
Add comprehensive JUnit tests

This commit adds a set of JUnit tests to enhance the test coverage of
the Duke application. The tests cover various aspects of the codebase,
including task management, parsing, commands, and error handling. Here's
a summary of the added tests:

1. TaskList Test - Adding and Deleting Tasks:
   - Verifies the functionality of adding and deleting tasks in the task
list.

2. TaskList Test - Marking Tasks as Done:
   - Ensures tasks can be marked as done in the task list.

3. Parser Test - Parsing Deadline Task:
   - Tests the parsing of deadline tasks from user input.

4. Parser Test - Invalid Date Format:
   - Checks that the parser handles invalid date formats correctly.

5. AddCommand Test - Adding a Deadline Task:
   - Verifies the behavior of adding a deadline task using the
AddCommand.

6. DeleteCommand Test - Deleting a Task:
   - Ensures the DeleteCommand correctly deletes tasks from the task
list.

7. MarkDoneCommand Test - Marking a Task as Done:
   - Validates the MarkDoneCommand's ability to mark tasks as done.

8. FindCommand Test - Finding Tasks by Keyword:
   - Tests the FindCommand's ability to find tasks by keyword.

9. PriorityComparator Test:
   - Checks the PriorityComparator's sorting behavior for tasks.

These tests aim to cover nearly all code that can be tested
automatically, improving the robustness and reliability of the Duke
application. Manual testing for GUI functionality and different
environment settings should complement these automated tests.

Please review and merge this commit to incorporate the new tests into the codebase.

AryanG01 committed:
Update Axela User Guide

The current user guide lacks detailed information on Axela's features
and usage, making it less helpful for users.

In this commit, I have added comprehensive content to the user guide to
ensure users can fully utilize Axela's capabilities. I've included
detailed explanations of features, usage instructions, and error
handling guidance. This enhances the user experience and provides
clarity on how to interact with Axela effectively.

This change aims to improve user understanding and make Axela a more
user-friendly AI assistant.

AryanG01 committed:
Creating and fixing jar file

AryanG01 committed:
Fixing Checstyle

AryanG01 committed:
Revert "Fixing Checstyle"

This reverts commit d45db9e4b15bfb02e935f7fdfe776195d3e9b30a.

AryanG01 committed:
Update Readme.md

AryanG01 committed:
commit changes
